### DISCUSSION TITLE
Concatenate

### DISCUSSION THREAD

**Discussion by valentinavan** [spectrochempy/spectrochempy#733] [Q&A] [answered]
Hi,

I am trying to concatenate a serie of NDDataset using `all_spectra = scp.concatenate(nds_list, axis='y') ` however it is impossible to track the order of the original NDDataset inside all_spectra unless i compare each single plot manually. 
I have tried multiple options to make it work but I was not successful:
```
for var_name in dir():
    # Check if the variable matches the pattern "s\d+"
    if re.match(r's\d+', var_name):
        nd_dataset = globals()[var_name]
        nd_dataset.x.title = f"{var_name.replace('_c', '')}"
```
I have tried with `nd_dataset.name = f"{var_name.replace('_c', '')}"` or `nd_dataset.dlabel= f"{var_name.replace('_c', '')}"` etc..but then when I concatenate it names all the individual NDDataset in `all_spectra` with the same name (the first or the last in `nds_list`)

How can I concatenate maintaining the name of each NDDataset so that I know which spectra is which?

Thanks

**Comment by atravert** [comment, marked as answer]
Hi @valentinavan ,

Thank you for your message. To indentify the individual spectra, you can use the label property, for instance:

```python
all_spectra = scp.concatenate([ds1, ds2], axis='y')
for spectrum_labels in ds1.y.labels:
    print(spectrum_labels)
print('---')
for spectrum_labels in s2.y.labels:
    print(spectrum_labels)
print('---')
for spectrum_labels in all_spectra.y.labels:
    print(lspectrum_labels)
```

The concatenation preserves the original order of the spectra, i.e. the first spectrum of `all_spectra` is the first spectrum of `ds1` and the last spectrum of `all_spectra` is the last of `ds2`.  

Note also that the `labels` attribute is protected, so that:
```python
s1.y.labels = [f"spectrum #{i}" for i in range(s1.shape[0])]
```
will **add** new labels to the existing ones,

```python
s1.y._labels = [f"spectrum {i}" for i in range(s1.shape[0])]
```
will **replace** the existing labels by the new ones,
 
I hope this helps. Let us know if this answers your question,

Best
